File: 2021_summer_algorithm_study_intermediate/using_cpp/6543_Tarjan.cpp
/*
input :
3 3
1 3 2 3 3 1
2 1
1 2
0

output :
1 3
2
*/

#include <iostream>
#include <algorithm>
#include <vector>
#include <cstring>
using namespace std;

// 전역 변수 선언
int V, E;
int SCC_NUM;
int NODE_NUM;
vector<int> graph[5005];
vector<int> st;
int scc[5005];
int dfn[5005];
int low[5005];
bool finished[5005];

bool comp(vector<int> a, vector<int> b){
    if (a.size() == 0) return true;
    if (b.size() == 0) return false;
    return a[0] < b[0];
}

void dfs(int s){
    dfn[s] = low[s] = ++NODE_NUM;
    st.push_back(s);

    for (int nxt: graph[s]){
        if (dfn[nxt] == 0){
            dfs(nxt);
            low[s] = min(low[s], low[nxt]);
        }
        else{
            if (finished[nxt]) continue;

            low[s] = min(low[s], dfn[nxt]);
        }
    }

    if (low[s] == dfn[s]){
        SCC_NUM++;
        while (true){
            int node = st.back(); st.pop_back();
            finished[node] = true;
            scc[node] = SCC_NUM;

            if (node == s) break;
        }
    }
}

bool sink_chk(vector<int> group){
    int group_scc = scc[group[0]];
    for (int node: group){
        for (int adj: graph[node]){
            if (scc[adj] != group_scc) return false;
        }
    }
    return true;
}

int main(){
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    freopen("input.txt", "r", stdin);

    while (true){
        for (auto& nodes: graph) nodes.clear();
        memset(scc, 0, sizeof(scc));
        memset(dfn, 0, sizeof(dfn));
        memset(low, 0, sizeof(low));
        memset(finished, false, sizeof(finished));
        SCC_NUM = 0; NODE_NUM = 0;

        cin >> V;
        if (V == 0) return 0;
        cin >> E;

        while(E--){
            int a, b;
            cin >> a >> b;
            graph[a].push_back(b);
        }

        for (int i = 0; i <= V; i++){
            if (finished[i]) continue;
            dfs(i);
        }

        // SCC 결과 생성
        vector<int> res[5005];
        for (int i = 1; i <= V; i++){
            res[scc[i]].push_back(i);
        }
        sort(res + 1, res + SCC_NUM + 1, comp);

        // 하나의 scc그룹이 outdegree가 없는지 체크하고, 없으면 ans에 넣음
        vector<int> ans;
        for (vector<int> group: res){
            if (group.size() == 0) continue;
            if (!sink_chk(group)) continue;

            for (int i: group) ans.push_back(i);
        }
        sort(ans.begin(), ans.end());

        for (int node: ans){
            cout << node << ' ';
        }
        cout << '\n';
    }
}

File: 2021_summer_algorithm_study_intermediate/using_python/6543_Tarjan.py
"""
input :
3 3
1 3 2 3 3 1
2 1
1 2
0

output :
1 3
2
"""
import sys
sys.stdin = open("input.txt", "r")
input = sys.stdin.readline


def dfs(s):
    global NODE_NUM; global SCC_NUM
    NODE_NUM += 1
    dfn[s] = low[s] = NODE_NUM
    st.append(s)

    for nxt in graph[s]:
        if dfn[nxt] == 0:  # 처음 방문. 즉, tree edge를 타고 방문
            dfs(nxt)
            low[s] = min(low[s], low[nxt])
        else:
            if finished[nxt]: continue
            low[s] = min(low[s], dfn[nxt])

    if dfn[s] == low[s]:
        SCC_NUM += 1
        while True:
            node = st.pop()
            finished[node] = True
            scc[node] = SCC_NUM

            if node == s: break


def sink_chk(group):
    group_scc = scc[group[0]]
    for node in group:
        for adj in graph[node]:
            if scc[adj] != group_scc:
                return False
    return True


if __name__ == "__main__":
    try:
        while True:
            V, E = map(int, input().rstrip().split())
            nodes = list(map(int, input().rstrip().split()))
            graph = [[] for _ in range(V + 1)]
            for i in range(0, 2 * E, 2):
                graph[nodes[i]].append(nodes[i + 1])

            SCC_NUM = 0
            NODE_NUM = 0
            dfn = [0 for _ in range(V + 1)]
            low = [0 for _ in range(V + 1)]
            finished = [False for _ in range(V + 1)]
            scc = [0 for _ in range(V + 1)]
            st = []
            for i in range(1, V + 1):
                # cross edge를 무시하기 위한 용도. pop을 한 바 있으면 finished된 것으로 처리
                if finished[i]: continue
                dfs(i)

            # SCC 결과 생성
            res = [[] for _ in range(V + 1)]
            for i in range(1, V + 1):
                res[scc[i]].append(i)
            res.sort()

            # 하나의 scc 그룹이 outdegree가 없는지 체크하고, 없으면 ans에 넣음
            ans = []
            for group in res:
                if len(group) == 0: continue
                if not sink_chk(group): continue

                ans += group

            ans.sort()
            for node in ans:
                print(node, end=' ')
            print()
    except:
        pass
